xpublish_edr/select.py

```diff
@@ -0,0 +1,125 @@
+                    f"Invalid datetimes submitted - {datetimes}",
+                )
+        except ValueError as e:
+            logger.error("Error with datetime", exc_info=True)
+            raise ValueError(f"Invalid datetime ({e})") from e
+
+    if query.parameters:
+        try:
+            ds = ds.cf[query.parameters.split(",")]
+        except KeyError as e:
+            raise ValueError(f"Invalid variable: {e}") from e
+
+    query_param_keys = list(query_params.keys())
+    for query_param in query_param_keys:
+        if query_param in edr_query_params:
+            del query_params[query_param]
+
+    method: Optional[str] = "nearest"
+
+    for key, value in query_params.items():
+        split_value = value.split("/")
+        if len(split_value) == 1:
+            continue
+        elif len(split_value) == 2:
+            query_params[key] = slice(split_value[0], split_value[1])
+            method = None
+        else:
+            raise ValueError(f"Too many values for selecting {key}")
+
+    ds = ds.sel(query_params, method=method)
+    return ds
+
+
+def select_postition(ds: xr.Dataset, point: Point) -> xr.Dataset:
+    """
+    Return a dataset with the position nearest to the given coordinates
+    """
+    if _is_regular_xy_coords(ds):
+        return _select_position_regular_xy_grid(ds, point)
+    else:
+        # TODO: Handle 2D coordinates
+        raise NotImplementedError("Only 1D coordinates are supported")
+
+
+def select_area(ds: xr.Dataset, polygon: Polygon) -> xr.Dataset:
+    """
+    Return a dataset with the area within the given polygon
+    """
+    if _is_regular_xy_coords(ds):
+        return _select_area_regular_xy_grid(ds, polygon)
+    else:
+        # TODO: Handle 2D coordinates
+        raise NotImplementedError("Only 1D coordinates are supported")
+
+
+def _coord_is_regular(da: xr.DataArray) -> bool:
+    """
+    Check if the DataArray has a regular grid
+    """
+    return len(da.shape) == 1 and da.name in da.dims
+
+
+def _is_regular_xy_coords(ds: xr.Dataset) -> bool:
+    """
+    Check if the dataset has 2D coordinates
+    """
+    return _coord_is_regular(ds.cf["X"]) and _coord_is_regular(ds.cf["Y"])
+
+
+def _select_position_regular_xy_grid(ds: xr.Dataset, point: Point) -> xr.Dataset:
+    """
+    Return a dataset with the position nearest to the given coordinates
+    """
+    # Find the nearest X and Y coordinates to the point
+    return ds.cf.sel(X=point.x, Y=point.y, method="nearest")
+
+
+def _select_area_regular_xy_grid(ds: xr.Dataset, polygon: Polygon) -> xr.Dataset:
+    """
+    Return a dataset with the area within the given polygon
+    """
+    # For a regular grid, we can create a meshgrid of the X and Y coordinates to create a spatial mask
+    pts = np.meshgrid(ds.cf["X"], ds.cf["Y"])
+
+    # Create a mask of the points within the polygon
+    contains = np.vectorize(lambda p: polygon.contains(Point(p)), signature="(n)->()")
+    mask = contains(np.stack(pts, axis=-1))
```

Comment:
```suggestion
    mask = shapely.contains_xy(polygon, pts[0], pts[1])
```
maybe?

I have learned that shapely has all kinds of useful functions that run the inner in C through GEOS, but are quite hard to find. This is potentially one of them: https://shapely.readthedocs.io/en/stable/reference/shapely.contains_xy.html#shapely.contains_xy

Reply:
[shapely.readthedocs.io/en/stable/reference/shapely.contains.html#shapely.contains](https://shapely.readthedocs.io/en/stable/reference/shapely.contains.html#shapely.contains):

"following this definition, a geometry does not contain its boundary, but it does contain itself. "

So... if you request a polygon where a vertex is on the grid, what do we get?

Reply:
It clips. What do we expect though? My intuition is that it should be included if it hits the boundary? My first thought is a tiny buffer?

Reply:
does the standard say anything?

Reply:
nope